### PR TITLE
Feature/back strafing steps

### PR DESCRIPTION
**Pull request by RogPodge**
## Overview
Cleanup and finishing an older PR: #9035

## Changes
- Fixes: #6706

### CODE REVIEW

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs [hunk 1]
```diff
@@ -85,7 +87,20 @@ public bool TeleportHotSpotCursorVisibility
 
         [SerializeField]
         [Tooltip("The distance to move the camera when the strafe is activated")]
-        private float strafeAmount = 0.25f;
+        internal float strafeAmount = 0.25f;
+        
+        [SerializeField]
+        [Tooltip("Whether or not a strafe checks that there is a floor beneath the user's origin on strafe")]
+        internal bool checkForFloorOnStrafe = default;
```
Comment (MaxWang-MS) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:94]: I am not sure if we prefer the explicit value or default?

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs [hunk 2]
```diff
@@ -231,6 +246,59 @@ protected Gradient GetLineGradient(TeleportSurfaceResult targetResult)
             }
         }
 
+        /// check if a backstrafe is possible on a valid platform regarding the possible strafe height given
```
Comment (MaxWang-MS) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:249]: Missing `<summary>`

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs [hunk 3]
```diff
@@ -231,6 +246,59 @@ protected Gradient GetLineGradient(TeleportSurfaceResult targetResult)
...
+            canMove = false;
+            var height = MixedRealityPlayspace.Position.y;
+            var newPosition = -CameraCache.Main.transform.forward * strafeAmount + MixedRealityPlayspace.Position;
+
+            newPosition.y = height;
+            bool isValidStrafe = true;
+            if (checkForFloorOnStrafe)
+            {
+                isValidStrafe = CheckPossibleBackStep(newPosition, out var strafeHitPosition);
+                if (adjustHeightOnStrafe)
```
Comment (MaxWang-MS) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:290]: Do we want to adjust the height even if the strafe is not valid?

Comment (RogPodge) [reply Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:290]: Good catch I'll fix this

Comment (RogPodge) [reply Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:290]: [Updated the comment after looking into this]

**Review by MaxWang-MS — APPROVED** [review]

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs [hunk 4]
```diff
@@ -231,6 +246,60 @@ protected Gradient GetLineGradient(TeleportSurfaceResult targetResult)
...
+            Vector3 strafeOrigin = new Vector3(newPosition.x, MixedRealityPlayspace.Position.y + maxHeightChangeOnStrafe, newPosition.z);
+            Vector3 strafeTerminus = strafeOrigin + (Vector3.down * maxHeightChangeOnStrafe * 2f);
+            Debug.Log("origin" + strafeOrigin);
+            Debug.Log("terminus" + strafeTerminus);
+
+            RayStep rayStep = new RayStep(strafeOrigin, strafeTerminus);
+            LayerMask[] layerMasks =  new LayerMask[] { ValidLayers };
+
+            // check are we hiting a floor plane or step above the current MixedRealityPlayspace.Position
+            if (raycastProvider.Raycast(rayStep, layerMasks, false, out var hitInfo))
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:267]: We may want to null check `raycastProvider`

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs [hunk 5]
```diff
@@ -231,6 +246,60 @@ protected Gradient GetLineGradient(TeleportSurfaceResult targetResult)
...
+        /// <param name="newPosition">the new position relative to backstrafe position</param>
+        /// <param name="hitStrafePosition">actual position the strafe raycast hits</param>
+        /// <returns>if there is a valid layer one can backstrafe on</returns>
+        internal bool CheckPossibleBackStep(Vector3 newPosition, out Vector3 hitStrafePosition)
+        {
+            var raycastProvider = CoreServices.InputSystem.RaycastProvider;
+            Vector3 strafeOrigin = new Vector3(newPosition.x, MixedRealityPlayspace.Position.y + maxHeightChangeOnStrafe, newPosition.z);
+            Vector3 strafeTerminus = strafeOrigin + (Vector3.down * maxHeightChangeOnStrafe * 2f);
+            Debug.Log("origin" + strafeOrigin);
+            Debug.Log("terminus" + strafeTerminus);
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:261]: Do we need these logs?

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs [hunk 6]
```diff
@@ -231,6 +246,60 @@ protected Gradient GetLineGradient(TeleportSurfaceResult targetResult)
...
+
+            RayStep rayStep = new RayStep(strafeOrigin, strafeTerminus);
+            LayerMask[] layerMasks =  new LayerMask[] { ValidLayers };
+
+            // check are we hiting a floor plane or step above the current MixedRealityPlayspace.Position
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:266]:
```suggestion
            // check are we hitting a floor plane or step above the current MixedRealityPlayspace.Position
```

**Review by keveleigh — APPROVED** [review]

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs (hunk 5 above)
Comment (RogPodge) [reply Assets/MRTK/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:261]: thanks for checking my excess log statements lol

**Pull request merged**